Source/Diagnostics/ReducedDiags/ParticleHistogram.cpp

```diff
@@ -143,46 +146,78 @@ void ParticleHistogram::ComputeDiags (int step)
 
-    for ( int i = 0; i < m_bin_num; ++i )
-    {
-        // compute the histogram
-        m_data[i] = ReduceSum( myspc,
-        [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+    // zero-out old data on the host
+    std::fill(m_data.begin(), m_data.end(), amrex::Real(0.0));
+    amrex::Gpu::DeviceVector< amrex::Real > d_data( m_data.size(), 0.0 );
+    amrex::Real* const AMREX_RESTRICT dptr_data = d_data.dataPtr();
+
+    int const nlevs = std::max(0, myspc.finestLevel()+1);
+    for (int lev = 0; lev < nlevs; ++lev) {
+#ifdef AMREX_USE_OMP
+#pragma omp parallel
+#endif
         {
-            // return 0 if the particle is filtered out
-            if (!fun_filterparser(p, amrex::RandomEngine())) {return 0.0_rt;}
-            // continue function if particle is not filtered out
-            auto const w  = p.rdata(PIdx::w);
-            amrex::ParticleReal x, y, z;
-            get_particle_position(p, x, y, z);
-            auto const ux = p.rdata(PIdx::ux)/PhysConst::c;
-            auto const uy = p.rdata(PIdx::uy)/PhysConst::c;
-            auto const uz = p.rdata(PIdx::uz)/PhysConst::c;
-            auto const f = fun_partparser(t,x,y,z,ux,uy,uz);
-            auto const f1 = bin_min + bin_size*i;
-            auto const f2 = bin_min + bin_size*(i+1);
-            if ( f > f1 && f < f2 ) {
-                if ( is_unity_particle_weight ) return 1.0_rt;
-                else return w;
-            } else return 0.0_rt;
-        });
+            for (WarpXParIter pti(myspc, lev); pti.isValid(); ++pti)
+            {
+                auto const GetPosition = GetParticlePosition(pti);
+
+                auto & attribs = pti.GetAttribs();
+                Real* const AMREX_RESTRICT d_w = attribs[PIdx::w].dataPtr();
+                Real* const AMREX_RESTRICT d_ux = attribs[PIdx::ux].dataPtr();
+                Real* const AMREX_RESTRICT d_uy = attribs[PIdx::uy].dataPtr();
+                Real* const AMREX_RESTRICT d_uz = attribs[PIdx::uz].dataPtr();
+
+                long const np = pti.numParticles();
+
+                //Flag particles that need to be copied if they cross the z_slice
+                amrex::ParallelFor(np,
+                   [=] AMREX_GPU_DEVICE(int i)
+                {
+                    amrex::ParticleReal x, y, z;
+                    GetPosition(i, x, y, z);
+                    auto const w  = d_w[i];
+                    auto const ux = d_ux[i] / PhysConst::c;
+                    auto const uy = d_uy[i] / PhysConst::c;
+                    auto const uz = d_uz[i] / PhysConst::c;
+
+                    // don't count a particle if it is filtered out
+                    if (!fun_filterparser(t, x, y, z, ux, uy, uz)) return;
+                    // continue function if particle is not filtered out
+                    auto const f = fun_partparser(t, x, y, z, ux, uy, uz);
+
+                    // determine particle bin
+                    int const bin = int(std::floor((f-bin_min)/bin_size));
```

Comment:
For DPC++ from `AMReX_Math.H`:
```suggestion
                    int const bin = int(Math::floor((f-bin_min)/bin_size));
```